--- src/pages/profiles/[userId].tsx
+++ src/pages/profiles/[userId].tsx
@@ -336,13 +336,11 @@
     <>
       <Heading size="md">生活照</Heading>
       <FormControl>
-        {profile.照片链接 && (
-          <Image
-            src={profile.照片链接}
-            alt="照片"
-            maxW="300px"
-            my={componentSpacing}
-          />
-        )}
+        <Image
+          src={UserProfilePictureLink(profile)}
+          alt="照片"
+          maxW="300px"
+          my={componentSpacing}
+        />

         {isMe && (
